Improve the Datadog scaler with non-breaking changes

arapulido opened the pull request:
Several non-breaking fixes for the Datadog scaler. Fixes #2657

The problem described in #2657 is partially caused by the selection of a time window that is too small (15 seconds). Datadog doesn't have that metric yet in many cases, and returns an empty value, that logs as a warning in the HPA.

To avoid this from happening often, the user should select a bigger window, and also added some retries before giving up. I also changed the documentation, to discourage reducing the time window: https://github.com/kedacore/keda-docs/pull/682

To avoid this, we should remove the "age" parameter going forward, but as this is a breaking change, I will create a different PR for 2.7.

Other improvements in this PR:

* If more than one point is returned, return the last one,  instead of the first one, for more recent results
* More parsing of the initial query, and some potential fixes to a incomplete query: default to avg aggregator, and only add a rollup if one wasn't added by the user
* Moved to MilliQuantities and floats for query results for more granularity
* Return a specific error if Datadog didn't return a metric value due to reaching rate limits for the organization
* IsActive doesn't try to get a metric value, to reduce rate limiting

### Checklist

- [X] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [X] Tests have been added
- [X] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

 Fixes #2657

JorTurFer commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -60,17 +61,53 @@ func NewDatadogScaler(ctx context.Context, config *ScalerConfig) (Scaler, error)
 
+// parseAndTransformDatadogQuery checks correctness of the user query and adds rollup if not available
+func parseAndTransformDatadogQuery(q string, age int) (string, error) {
+	aggregator := regexp.MustCompile(`^(avg|sum|min|max):.*`)
```
Maybe, this regex could be moved at package level to avoid compilation on every call.
You can find an example [here](https://github.com/kedacore/keda/blob/main/pkg/scalers/rabbitmq_scaler.go#L25-L29)

WDYT?

arapulido replied:
👍

JorTurFer commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -60,17 +61,53 @@ func NewDatadogScaler(ctx context.Context, config *ScalerConfig) (Scaler, error)
+		q = "avg:" + q
+	}
+
+	filter := regexp.MustCompile(`.*\{.*\}.*`)
```
The same with this regex

arapulido replied:
👍

JorTurFer commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -60,17 +61,53 @@ func NewDatadogScaler(ctx context.Context, config *ScalerConfig) (Scaler, error)
+		return "", fmt.Errorf("malformed Datadog query")
+	}
+
+	rollup := regexp.MustCompile(`.*\.rollup\(.*\)`)
```
This regex too

arapulido replied:
👍

JorTurFer commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -174,8 +197,12 @@ func (s *datadogScaler) Close(context.Context) error {
 
-// IsActive returns true if we are able to get metrics from Datadog
 func (s *datadogScaler) IsActive(ctx context.Context) (bool, error) {
+	return true, nil
```
Maybe I'm wrong (correct me in that case @zroubalik ), but if the scaler returns `true` always in this function, this scaler can't scale to zero. 
Is this intended? We should add a note in docs in that case

zroubalik replied:
Yeah, it shouldn't behave like that. THe `IsActive` should mark the activity of the scaler - ie whether the target vaules in Datadog are above (Active) or below (Not Active) the threshold.

JorTurFer commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -195,66 +222,68 @@ func (s *datadogScaler) IsActive(ctx context.Context) (bool, error) {
 
-	points := series[0].GetPointlist()
+		resp, r, err := s.apiClient.MetricsApi.QueryMetrics(ctx, time.Now().Unix()-int64(s.metadata.age), time.Now().Unix(), s.metadata.query)
 
-	if len(points) == 0 {
-		return false, nil
-	}
+		if r.StatusCode == 429 {
+			rateLimit := r.Header.Get("X-Ratelimit-Limit")
+			rateLimitReset := r.Header.Get("X-Ratelimit-Reset")
 
-	return true, nil
-}
+			return -1, fmt.Errorf("your Datadog account reached the %s queries per hour rate limit, next limit reset will happen in %s seconds: %s", rateLimit, rateLimitReset, err)
+		}
```
I wouldn't do this. HPA controller runs in single thread and with this we are blocking it for up to 25 seconds. This could produce impact in the whole cluster

zroubalik replied:
Yeah, the call should be definitely non blocking.

JorTurFer reviewed (changes requested):
Thanks for your contribution ❤️ ❤️ 
I have left some comments

arapulido commented:
@tomkerkhove I implemented here the `unavailableMetricValue` optional parameter. 

@JorTurFer I addressed your comments

Once this is approved, I will create a PR with the updated documentation

The missing thing is `IsActive`. I am not sure what needs implementing here. Anything that requires contacting the API will accelerate rate limiting, and it is not reliable. Any suggestions?

JorTurFer commented:
> @tomkerkhove I implemented here the `unavailableMetricValue` optional parameter.
> 
> @JorTurFer I addressed your comments
> 
> Once this is approved, I will create a PR with the updated documentation
> 
> The missing thing is `IsActive`. I am not sure what needs implementing here. Anything that requires contacting the API will accelerate rate limiting, and it is not reliable. Any suggestions?

The point here is that KEDA uses that method for getting if the scaler is active or not and that triggers things like scale from/to zero. If we return true always, KEDA thinks that the scaler is always active and will not scale to zero

arapulido commented:
> The point here is that KEDA uses that method for getting if the scaler is active or not and that triggers things like scale from/to zero. If we return true always, KEDA thinks that the scaler is always active and will not scale to zero

My question is: what do we mean by "a scaler is active"?

JorTurFer commented:
I'm AFK but this is part of the info (from docs) 

> KEDA polls ScaledObject object according to the pollingInterval configured in the ScaledObject; it checks the last time it was polled, it checks if the number of replicas is greater than 0, and if the scaler itself is active. So if the scaler returns false for IsActive, and if current number of replicas is greater than 0, and there is no configured minimum pods, then KEDA scales down to 0.

JorTurFer commented:
So it could be like:
IsActive=true: KEDA should scale from zero and let the HPA works
IsActive=false: KEDA should scale to zero (HPA doesn't work)

zroubalik commented:
> > The point here is that KEDA uses that method for getting if the scaler is active or not and that triggers things like scale from/to zero. If we return true always, KEDA thinks that the scaler is always active and will not scale to zero
> 
> My question is: what do we mean by "a scaler is active"?

As I mentioned above: https://github.com/kedacore/keda/pull/2694#discussion_r817483431

Active is when the target metric is above threshold. In this case a the value returned by `query` is above `queryValue`.

arapulido commented:
@JorTurFer @zroubalik Thanks both! I have finally understood this :) I have added a commit for this, and I have also opened a PR to the Datadog scaler docs to improve the explanation related to polling intervals configuration, as it is fairly important to understand it, as it is directly affected by rate limits: https://github.com/kedacore/keda-docs/pull/702

JorTurFer commented:
/run-e2e datadog*

JorTurFer reviewed (approved):
LGTM!
Thanks for these improvements!
(I'll wait to merge till e2e passes and docs PR is ready to merge too)

JorTurFer commented:
e2e test has failed and also the workflow has failed because a needed rebase (rebasing main is not important at this moment if you don't want to do it, e2e test runs and only fails after it trying to clean up all namespaces).
This is [the link](https://github.com/kedacore/keda/runs/5464458566?check_suite_focus=true#step:7:116) to datadog e2e logs

arapulido commented:
> e2e test has failed and also the workflow has failed because a needed rebase (rebasing main is not important at this moment if you don't want to do it, e2e test runs and only fails after it trying to clean up all namespaces). This is [the link](https://github.com/kedacore/keda/runs/5464458566?check_suite_focus=true#step:7:116) to datadog e2e logs

@JorTurFer  Added a couple of changes to make the tests more resilient. Can we rerun, please?

JorTurFer commented:
/run-e2e datadog*

JorTurFer commented:
They have worked (again, ignore the error, it's because a rebase is needed for fixing it)
Have you got any other thing to add in this PR? 
Could you update the PR with the issue please? Then I can merge this

As always, thanks a ton for you nice work ❤️ ❤️ ❤️ ❤️

arapulido commented:
@JorTurFer Thanks! I don't have plans to add anything for now, but I need to know whether this is going to be targeted for `2.6` or `2.7`, as I need to add changes to the documentation related to the new `metricUnavailableValue` parameter

JorTurFer commented:
This change will be shipped as part of `2.7`

arapulido commented:
> This change will be shipped as part of `2.7`

Cool! The corresponding docs PR is here: https://github.com/kedacore/keda-docs/pull/710

zroubalik reviewed (approved):
LGTM

zroubalik commented:
@arapulido could you please update the Changelog? then we can merge this one. thanks!

zroubalik commented on CHANGELOG.md:
```diff
@@ -37,6 +37,7 @@
 - **Azure Queue:** Don't call Azure queue GetProperties API unnecessarily ([#2613](https://github.com/kedacore/keda/pull/2613))
 - **Kafka Scaler** Make "disable" a valid value for tls auth parameter ([#2608](https://github.com/kedacore/keda/issues/2608))
 - **Datadog Scaler:** Validate query to contain `{` to prevent panic on invalid query ([#2625](https://github.com/kedacore/keda/issues/2625))
+- **Datadog Scaler:** Several improvements, including a new optional parameter `metricUnavailableValue` to fill data when no Datadog metric was returned ([#2694](https://github.com/kedacore/keda/pull/2694))
```
Could you please refer here to the issue, not pull request?

zroubalik reviewed (changes requested):
Could you please rebase, there's a conflict in Changelog. Thanks!

JorTurFer commented on CHANGELOG.md:
```diff
@@ -39,6 +39,7 @@
 - **Datadog Scaler:** Validate query to contain `{` to prevent panic on invalid query ([#2625](https://github.com/kedacore/keda/issues/2625))
 - **Kafka Scaler** Make "disable" a valid value for tls auth parameter ([#2608](https://github.com/kedacore/keda/issues/2608))
 - **RabbitMQ Scaler:** Include `vhost` for RabbitMQ when retrieving queue info with `useRegex` ([#2498](https://github.com/kedacore/keda/issues/2498))
+- **Datadog Scaler:** Several improvements, including a new optional parameter `metricUnavailableValue` to fill data when no Datadog metric was returned ([#2657](https://github.com/kedacore/keda/issues/2657))
```
Sorry for being a pain, but please move up this for keeping record alphabetically sorted

arapulido replied:
Done!

zroubalik commented on pkg/scalers/datadog_scaler.go:
```diff
@@ -60,17 +71,61 @@ func NewDatadogScaler(ctx context.Context, config *ScalerConfig) (Scaler, error)
+		return "", fmt.Errorf("malformed Datadog query")
+	}
+
+	match := rollup.FindStringSubmatch(q)
```
Would be nice to add some comment here, with an example contents that are being parsed and some description. Just for clarity.

arapulido replied:
Added comments and added tests for that function

zroubalik reviewed (approved):
LGTM, thanks!

zroubalik commented:
/run-e2e datadog*
